pkg/controller/util_test.go

```diff
@@ -0,0 +1,175 @@
+package controller
+
+import (
+	"context"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller"
+)
+
+// newRefGrantList returns a new ReferenceGrant with given attributes
+func newRefGrantList(name, fromNamespace, fromGrop, fromKind, toNamespace, toGroup, toKind, toName string, settoName bool) []*gatewayv1beta1.ReferenceGrant {
+	ReferenceGrantList := []*gatewayv1beta1.ReferenceGrant{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: fromNamespace,
+			},
+			Spec: gatewayv1beta1.ReferenceGrantSpec{
+				From: []gatewayv1beta1.ReferenceGrantFrom{
+					{
+						Group:     gatewayv1beta1.Group(fromGrop),
+						Kind:      gatewayv1beta1.Kind(fromKind),
+						Namespace: gatewayv1beta1.Namespace(toNamespace),
+					},
+				},
+				To: []gatewayv1beta1.ReferenceGrantTo{
+					{
+						Group: gatewayv1beta1.Group(toGroup),
+						Kind:  gatewayv1beta1.Kind(toKind),
+					},
+				},
+			},
+		},
+	}
+
+	if settoName {
+		objectName := gatewayv1beta1.ObjectName(toName)
+		ReferenceGrantList[0].Spec.To[0].Name = &objectName
+	}
+
+	return ReferenceGrantList
+}
+
+func TestIsGranted(t *testing.T) {
+	ctx := context.Background()
+	var requestedBytes int64 = 1000
+	snapapiGrp := "snapshot.storage.k8s.io"
+	snapapiKind := "VolumeSnapshot"
+	coreapiGrp := ""
+	pvcapiKind := "PersistentVolumeClaim"
+	anyvolumedatasourceapiGrp := "hello.k8s.io/v1alpha1"
+	anyvolumedatasourceapiKind := "Hello"
+	fromNamespace := "ns1"
+	fromsrcName := "test-dataSource"
+	toNamespace := "ns2"
+
+	type testcase struct {
+		expectErr    bool
+		volOpts      controller.ProvisionOptions
+		refGrantList []*gatewayv1beta1.ReferenceGrant
+	}
+	testcases := map[string]testcase{
```

Comment:
Can you add a test case for nil datasource apigroup?

Reply:
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-640957054096359bf678ef9f2789673a60814fba8286165ad298f5eba881536aR215-R222
I added.